akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala

```diff
@@ -63,26 +65,23 @@ import akka.projection.StatusObserver
       if (abort.isCompleted) {
         abort
       } else {
-        try {
+        (try {
           futureCallback()
         } catch {
           case NonFatal(e) =>
             // in case the callback throws instead of returning failed Future
             Future.failed(e)
+        }).recoverWith {
+          // using recoverWith instead of `.failed.foreach` to make sure that calls to statusObserver
+          // are invoked in sequential order
+          case err if !abort.isCompleted =>
+            telemetry.error(projectionId, err)
+            statusObserver.error(projectionId, env, err, recoveryStrategy)
+            Future.failed(err)
         }
       }
     }
 
-    val retryFutureCallback: () => Future[Done] = { () =>
-      tryFutureCallback().recoverWith {
-        // using recoverWith instead of `.failed.foreach` to make sure that calls to statusObserver
-        // are invoked in sequential order
-        case err if !abort.isCompleted =>
-          statusObserver.error(projectionId, env, err, recoveryStrategy)
-          Future.failed(err)
-      }
-    }
```

Comment:
@patriknw this PR is still DRAFT but I noticed a change that might be useful to bring `StatusObserver` and `Telemetry` into a smaller code block (instead of being a bit here and a bit there).

Now that we have `InternalProjectionState` I think we can take it out `statusObserver` away from `HandlerRecoveryImpl` and up into `InternalProjectionState`. In `HandlerRecoveryImpl`, `statusObserver.error` is invoked in a couple of places but it can be simplified. Then, we could add the `recoverWith` block on every call site to `handlerRecovery.applyRecovery` and take StatusObserved away from `HandlerRecoveryImpl`.

I'm not pushing forward that change because it requires some work on the `HandlerRecoveryImplSpec` (remove all the assertion wrt `StatusObserved` and rewrite those tests).